### PR TITLE
Routines using hardware sweepers

### PR DESCRIPTION
**Pull request by stavros11**
This is a prototype for using hardware sweepers to speed-up characterization scans, following the idea described in #187. It is a breaking change in the API so I would suggest to not merge it until the abstract platform and sweeper interface is decided. In the meantime I am using it to test the Quantum Machines driver from qiboteam/qibolab#259. 

This branch only works with qiboteam/qibolab#259. Here are some reports for the qw5q_gold using the QM instruments (needs access to the qq-upload server):

* [resonator punchout](http://login.qrccluster.com:9000/tYFmq18RRTSTajsXqKZpbQ==/)
* [resonator flux](http://login.qrccluster.com:9000/p3Uo2HnSS52eTON2MFussA==/)
* [qubit spectroscopy](http://login.qrccluster.com:9000/Yhn2vEkuRxaTF-WA3eHOyA==/)

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

### CODE REVIEW

**Comment by stavros11** [comment]
I merged the latest main into this. @andrea-pasquale while trying to run the resonator spectroscopy with this (using qiboteam/qibolab#259) I get the following error
```
builders.py 149 execute
self.update_platform_runcard(qubit, action)

builders.py 172 update_platform_runcard
params = data_fit.df[data_fit.df["qubit"] == qubit].to_dict("index")[0]

KeyError:
0
srun: error: zeraa2: task 0: Exited with exit code 1
```
which seems to be related to the fitting, not execution. Do you have any idea what that is? I was not getting it before refactoring 1.

**Comment by andrea-pasquale** [comment]
> I merged the latest main into this. @andrea-pasquale while trying to run the resonator spectroscopy with this (using [qiboteam/qibolab#259](https://github.com/qiboteam/qibolab/pull/259)) I get the following error
> 
> ```
> builders.py 149 execute
> self.update_platform_runcard(qubit, action)
> 
> builders.py 172 update_platform_runcard
> params = data_fit.df[data_fit.df["qubit"] == qubit].to_dict("index")[0]
> 
> KeyError:
> 0
> srun: error: zeraa2: task 0: Exited with exit code 1
> ```
> 
> which seems to be related to the fitting, not execution. Do you have any idea what that is? I was not getting it before refactoring 1.

Thanks for pointing this out @stavros11. I was not able to perform tests regarding multiplex within the first refactor so there may be some bugs. I fixed this one in #219.

File: src/qibocal/calibrations/characterization/resonator_spectroscopy.py [hunk 1]
```diff
@@ -87,12 +91,31 @@ def resonator_spectroscopy(
...
+        while (
+            any(result.in_progress for result in results.values())
+            or len(fast_sweep_data) == 0
+        ):
+            # retrieve the results for every qubit
+            for qubit in qubits:
+                # average msr, phase, i and q over the number of shots defined in the runcard
+                result = results[ro_pulses[qubit].serial]
+                # store the results
+                r = result.to_dict()
```
Comment (andrea-pasquale) [inline src/qibocal/calibrations/characterization/resonator_spectroscopy.py:106]: The code will fail again if you don't specify the `average` is `False`.
By default `to_dict` performs the average for i and q.
```suggestion
                r = result.to_dict(average=False)
```

**Comment by stavros11** [comment]
Following the merge of #234 here this PR should be ready for review. Some of the characterization routines were updated to use the new `Sweeper` object implemented in qibolab. These routines were tested by me on QM and @andrea-pasquale on qblox when the qubits were alive. Minor modifications were done since then but hopefully the code still works.

This PR also contains #175 and #203 as they were used to characterize two-qubit gates with QM. #175 (Chevron + tune landscape) was fully tested, although enhancements (eg. CZ using two flux pulses) can still be made. For #203 we could acquire data but we did not test the plotting and post-processing.

Note that when using hardware sweepers we only dump data at the end of the sweep, so qq-live won't be available for all routines that use sweepers.

**Review by andrea-pasquale — APPROVED** [review]

**Review by Edoardo-Pedicillo — APPROVED** [review]

**Pull request merged**